lib/src/cluster.dart

```diff
@@ -96,11 +98,79 @@ class Cluster implements Client {
     await for (final row in peers.asStream()) {
       final newIP = row.values[0] as InternetAddress;
       final oldPeers = _peers.map((peer) => peer.host);
-      if (oldPeers.contains(newIP)) ;
+      if (oldPeers.contains(newIP)) continue;
       final newPeer =
           await _Peer.connect(newIP, 9042, authenticator: _authenticator);
       _peers.add(newPeer);
     }
+    _loadPeerTokens();
+  }
+
+  Future<ResultPage> _queryPeer(
+      _Peer peer,
+      String query, {
+        Consistency consistency,
+        /* List | Map */
+        values,
+        int pageSize,
+        Uint8List pagingState,
+      }) {
+    consistency ??= _consistency;
+    final q = _Query(query, consistency, values, pageSize, pagingState);
+    final body = buildQuery(
+      query: query,
+      consistency: consistency,
+      values: values,
+      pageSize: pageSize,
+      pagingState: pagingState,
+    );
+    return peer._sendQuery(this, q, body);
+  }
+
+  _loadPeerTokens() async {
+    for (final peer in _peers) {
+      final tokenPage = _queryPeer(peer, 'SELECT tokens FROM system.local');
+      final tokenFirstPage = await tokenPage.asStream().first;
+      final tokenData = tokenFirstPage.items[0].values[0];
+      final tokens =
+          (tokenData as Set).map((token) => int.parse(token as String)).toSet();
+      _peerTokens[peer] = tokens;
+    }
+  }
+
+  _Peer _selectTokenPeer(String hint) {
+    final hash = murmur3_hash(hint);
+    _Peer bestPeer;
+    int bestToken;
+    for(final peer in _peers) {
+      // Selects the smallest token which is larger than the hash.
+      final closeToken = _peerTokens[peer].reduce((a, b) {
+        if(a < hash) {
+          return b;
+        }
+        if(b < hash) {
+          return a;
+        }
+        return min(a, b);
+      });
+      if(bestToken == null || closeToken < bestToken) {
+        bestToken = closeToken;
+        bestPeer = peer;
+      }
+    }
+    return bestPeer;
+  }
+
+  Future executeHint(
```

Comment:
Instead of `executeHint`, let's put the `hint` as part of `execute()` method.

```dart
  @override
  Future execute(
    String query, {
    Consistency consistency,
    /* List | Map */
    values,
    String hint,
  }) {
    consistency ??= _consistency;
    final peer = hint == null ? _selectPeer() : _selectTokenPeer(hint);
    return peer._sendExecute(query, consistency, values);
  }
```

Also do a similar change for `Future<ResultPage> query()`.